File: 백준/Silver/2583. 영역 구하기/영역 구하기.cc
#include <bits/stdc++.h>
using namespace std;
const int dx[] = {0, 1, 0, -1};
const int dy[] = {-1, 0 , 1, 0};
int m,n,k,sx,sy,ex,ey;
int a[101][101], visited[101][101];

int dfs(int y, int x){
	visited[y][x] = 1;
	int cnt = 1;
	for (int i=0; i < 4; i++){
		int nx = x + dx[i];
		int ny = y + dy[i];
		if (nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
		if (visited[ny][nx] == 1) continue;
		if (a[ny][nx]) continue;
		cnt += dfs(ny, nx);
	}
	return cnt;
}

int main(){
	vector<int> v;
	cin >> m >> n >> k;

	for (int i=0; i<k; i++){
		cin >> sx >> sy >> ex >> ey;
		for (int j=sy; j < ey; j++){
			for (int k = sx; k < ex; k++){
				a[j][k] = 1;
			}
		}
	}

	for(int i=0; i<m; i++){
		for (int j=0; j <n; j++){
			if (!visited[i][j] && a[i][j]!=1){
				v.push_back(dfs(i,j));
			}
		}
	}
	cout << v.size() << "\n";
	sort(v.begin(), v.end());
	for (int i : v)
		cout << i << " ";

	return 0;
}

File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

m,n,k = map(int, input().split())
dy=[-1,0,1,0]
dx=[0,1,0,-1]
mp = [[0] * n for _ in range(m)]

def paint(x1,y1,x2,y2):
    for i in range(y1, y2):
        for j in range(x1, x2):
            mp[i][j] = 1

for _ in range(k):
    s_x, s_y, e_x, e_y = map(int, input().split())
    paint(s_x, s_y, e_x, e_y)

def go(y,x):
    visited[y][x] = 1
    area = 1
    
    for i in range(4):
        ny=y+dy[i]
        nx=x+dx[i]
        if 0 <= nx < n and 0 <= ny < m:
            if visited[ny][nx] == 0 and mp[ny][nx] == 0:
                area += go(ny, nx)
    return area

visited = [[0]*n for _ in range(m)]
result = []
cnt = 0
for i in range(m):
    for j in range(n):
        if visited[i][j] == 0 and mp[i][j] == 0:
            cnt += 1
            result.append(go(i,j))

result.sort()
print(cnt)
print(*result)
